### PR TITLE
A mechanism for user options, and client side colourblind fix

### PR DESCRIPTION
**Pull request by TimothyJones**
Adds `objects/useroptions.php`, an object for handling user options. User options are stored in their own table, which is only populated when users save options (so it is lazily created for existing members). They are loaded when a `User` object is created. In addition to being part of the user object, options are also exposed to the client through `useroptions.php`, which produces a json object that describes a user's preferences for use in javascript.

The colorblindness fix is applied client side, depending on the user's options. It is triggered when the map changes, and replaces the map image with an appropriately adjusted version. Additionally, press is displayed with the country's name at the start of the message. Thanks to @Sleepcap for the suggestions and inspiration of this feature.

Additionally, the options contain an option to always hide the move arrows on the board, and to hide the new live game announcement on the home page.

Feedback welcome.

### CODE REVIEW

File: board/chatbox.php [hunk 1]
```diff
@@ -421,6 +421,15 @@ public function renderMessages($msgCountryID, $messages)
 			if( is_object($Member) && $message['fromCountryID'] == $Member->countryID )
 				$message['message'] = '<span class="messageFromMe">'.$message['message'].'</span>';
 
+			// Display the country name in front of the text (for colorblind people)
+			if ( $User->options->value['colourblind'] != 'Yes')
```
Comment (thorfi) [inline board/chatbox.php]: if ( $User->options->value['colourblind'] != 'No') // ?

File: objects/useroptions.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		$update = implode(',',$updates);		
+
+		$row = $DB->sql_hash("SELECT * FROM wD_UserOptions WHERE userID=".$this->id );
+		if ( ! isset($row['userID']) or ! $row['userID'] ) {
+			// create
+			$DB->sql_put("INSERT INTO wD_UserOptions (userID) VALUES (".$this->id.")");
+		} 
+		
+		$DB->sql_put("UPDATE wD_UserOptions SET $update WHERE userID=".$this->id);
+	}
```
Comment (thorfi) [inline objects/useroptions.php]: Is $DB auto committing, is there a commit somewhere else, or should there be one here?

File: locales/English/user.php [hunk 3]
```diff
@@ -86,7 +86,19 @@
...
-<?php
+
+<?php 
+	foreach ($User->options->value as $name=>$val) 
+	{
+		print '<li class="formlisttitle">'.UserOptions::$titles[$name].':</li>';
+		print '<li class="formlistdesc">';
+		foreach (UserOptions::$possibleValues[$name] as $possible) 
+		{
+			print ' <input type="radio" name="userForm['.$name.']" value="'.$possible.'" '. ($val == $possible ? 'checked' :'') . ' > '. $possible;
```
Comment (thorfi) [inline locales/English/user.php]: Not sure if you want to have a title="" element to explain the meaning of values (particularly the colour blind ones)?

File: javascript/mapUI.js [hunk 4]
```diff
@@ -137,11 +137,23 @@ function loadMap(gameID, currentTurn, newTurn)
 	
 	// Update the link to the large map
 	$('LargeMapLink').innerHTML = 
-			' <a href="map.php?gameID='+gameID+'&turn='+newTurn+'&mapType=large" target="blank" class="light">'+
+			' <a href="map.php?gameID='+gameID+'&turn='+newTurn+'&mapType=large'+(useroptions.showMoves =='No'?'&hideMoves':'')+'" target="blank" class="light">'+
 			'<img src="'+l_s('images/historyicons/external.png')+'" alt="'+l_t('Open large map')+'" ' +
 			'title="'+l_t('This button will open the large map in a new window. The large ' +
 			'map shows all the moves, and is useful when the small map isn\'t clear enough.')+'" /><\/a>';
```
Comment (thorfi) [inline javascript/mapUI.js]: title text might need fixing for the large map if you are hiding moves on the large map also?

File: board/chatbox.php (hunk 1 above)
Comment (TimothyJones) [inline board/chatbox.php]: Good catch. Fixed.

File: objects/useroptions.php (hunk 2 above)
Comment (kestasjk) [inline objects/useroptions.php]: Data is committed once the page has finished being generated (assuming no errors occur), but note that MyISAM tables aren't covered by transactions and they auto commit, only the innodb tables are protected by transactions.

**Pull request merged**